### PR TITLE
New executor

### LINKED ISSUE #1336: SynchronousExecutor for other dask schedulers
For running many lightweight IO-bound tasks, I've been using a custom threaded executor I hacked together:
```
class ThreadedExecutor(SynchronousExecutor):
    """Variant of built-in SynchronousExecutor that uses dask threaded scheduler."""

    @contextmanager
    def start(self) -> Iterator:
        """
        Context manager for initializing execution.

        Configures `dask` and yields the `dask.config` contextmanager.
        """
        with dask.config.set(scheduler="threads") as cfg:
            yield cfg

    def wait(self, futures: Any) -> Any:
        """
        Resolves a `dask.delayed` object to its values. Blocks until the computation is complete.

        Args:
            - futures (Any): iterable of `dask.delayed` objects to compute

        Returns:
            - Any: an iterable of resolved futures
        """
        with dask.config.set(scheduler="threads"):
            return dask.compute(futures)[0]
```
I'm not sure this warrants its own executor in Prefect, but perhaps it could be a configurable parameter? Being able to easily swap out the sync/threaded/multiprocess/distributed schedulers is a huge plus of dask, and right now Prefect supports is capitalizing on some but not all of that flexibility.

My own needs are already met by this little hack, just thought I'd contribute my $0.02.

### PR DESCRIPTION
**Pull request by cicdw**
**Thanks for contributing to Prefect!**

Please describe your work and make sure your PR:

- [x] adds new tests (if appropriate)
- [x] updates `CHANGELOG.md` (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

Note that your PR will not be reviewed unless all three boxes are checked.

## What does this PR change?
This PR implements a new `LocalDaskExecutor` for running Flows with various dask schedulers.  In addition, now that this executor supersedes `SynchronousExecutor` (and is easier to type), we will be deprecating the `SynchronousExecutor`.

## Why is this PR important?
Closes #1336 and allows for more fine-tuning of how your workflows should be run.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -39,7 +40,7 @@ These changes are available in the [master branch](https://github.com/PrefectHQ/
 
 ### Deprecations
 
-- None
+- Deprecate `SynchronousExecutor` in favor of `LocalDaskExecutor` - [#1434](https://github.com/PrefectHQ/prefect/pull/1434)
```
Comment (jlowin) [inline CHANGELOG.md]:
```suggestion
- Rename `SynchronousExecutor` as `LocalDaskExecutor` - [#1434](https://github.com/PrefectHQ/prefect/pull/1434)
```

File: src/prefect/config.toml [hunk 2]
```diff
@@ -71,7 +71,7 @@ checkpointing = false
     [engine.executor]
 
     # the default executor, specified using a full path
-    default_class = "prefect.engine.executors.SynchronousExecutor"
+    default_class = "prefect.engine.executors.LocalExecutor"
```
Comment (jlowin) [inline src/prefect/config.toml]: Out of curiosity, why this change?

File: src/prefect/engine/executors/dask.py [hunk 3]
```diff
@@ -201,3 +202,77 @@ def wait(self, futures: Any) -> Any:
...
+    def start(self) -> Iterator:
+        """
+        Context manager for initializing execution.
+
+        Configures `dask` and yields the `dask.config` contextmanager.
+        """
+        with dask.config.set(scheduler=self.scheduler, **self.kwargs) as cfg:
+            yield cfg
+
+    def queue(self, maxsize: int = 0) -> Queue:
```
Comment (jlowin) [inline src/prefect/engine/executors/dask.py]: Not important for this PR, but are we using these queues anywhere? Can we kill them?

File: CHANGELOG.md (hunk 1 above)
Comment (jlowin) [reply CHANGELOG.md]: (This may or may not be accurate)

File: src/prefect/config.toml (hunk 2 above)
Comment (cicdw) [reply src/prefect/config.toml]: I'm willing to revert this, but with the known possibility for double execution in deep mapped pipelines, this felt safer.  I actually _thought_ we had already done this

File: src/prefect/engine/executors/dask.py (hunk 3 above)
Comment (cicdw) [reply src/prefect/engine/executors/dask.py]: I can kill them in this PR, we are not using them anywhere!

Comment (cicdw) [reply src/prefect/engine/executors/dask.py]: Although maybe a user is? I'll add this to breaking changes...

**Comment by cicdw** [comment]
And yes, @jlowin this doesn't test the threaded executor.  I could update all of our executor tests to parametrize over this if you'd like.  My reasoning for it being "OK" (but maybe not 100% ideal) is that it's only changing dask's internal scheduling mechanism, and the fully distributed Dask executor tests for threading / process problems that could arise, so I didn't feel too uncomfortable with it.

File: src/prefect/engine/executors/dask.py (hunk 3 above)
Comment (cicdw) [reply src/prefect/engine/executors/dask.py]: Removed.

File: src/prefect/config.toml (hunk 2 above)
Comment (jlowin) [reply src/prefect/config.toml]: Totally fine, not asking for reversion was just curious why
